### PR TITLE
Payment coin and coin type

### PR DESCRIPTION
**Pull request by ghost**
This PR completes #962 

Features:

- Add `paymentCoin` to orders
- Add `coinType` and `paymentCoin` to cases/sales/purchases db tables
- Set `coinType` and `paymentCoin` when creating cases/sales/purchases db entries
- Migrate existing cases/sales/purchases db entries to set `coinType` and `paymentCoin`
- Migrate existing listing indexes to include `acceptedCurrencies`

### CODE REVIEW

**Comment by cpacia** [comment]
@tyler-smith where is the fix for the migration 09 -> 10 bug? Not sure if I see it.

**Comment by placer14** [comment]
@cpacia The migration test for 10 was fixed prior to this. See https://github.com/OpenBazaar/openbazaar-go/commit/9fb3fb72d4a70986136557d56f7fffbf19bbc27f

**Comment by cpacia** [comment]
got it

**Review by cpacia — APPROVED** [review]
I think this looks good to me. @placer14 can you also review?

File: repo/migrations/migration_helpers.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package migrations
+
+import (
+	"database/sql"
+	"os"
+	"path"
+	"strconv"
+)
+
+func NewDB(repoPath string, dbPassword string, testnet bool) (*sql.DB, error) {
```
Comment (placer14) [inline repo/migrations/migration_helpers.go]: Maybe more like `OpenDB`? Was surprised by the name and that it lived in `migrations` until I came to look.

Comment (placer14) [reply repo/migrations/migration_helpers.go]: One thing to be concerned with the idea of helpers in our migrations will be the problem that arises when the helper breaks across schema versions (if we were to move the location of the database file, for example). I tend to inline the helpers needed for each migration and write new ones. Something like:

`repo/migrations/migration_009.go`
```
func migration009_openDB (...) (...) {
   // actual open logic here
}
```

`repo/migrations/migration_010.go`
```
func migration010_openDB (...) (...) {
   return migration009_openDB(...)
}

// or explicitly show how this schema is changing 
// by manipulating the inputs as needed for migration009_openDB
func migration010_openDB (path string, ...) (...) {
   migration010_dbPath := migration010_dbPathChanges(path)
   return migration009_openDB(migration010_dbPath, ...)
}

```

File: test/factory/contract.go [hunk 2]
```diff
@@ -3,9 +3,10 @@ package factory
 import (
 	"time"
 
+	crypto "gx/ipfs/QmaPbCnUMBohSGo3KnxEa2bHqyJVVeEEcwtqJAYxerieBo/go-libp2p-crypto"
+
 	"github.com/OpenBazaar/openbazaar-go/pb"
 	"github.com/golang/protobuf/ptypes"
-	crypto "gx/ipfs/QmaPbCnUMBohSGo3KnxEa2bHqyJVVeEEcwtqJAYxerieBo/go-libp2p-crypto"
```
Comment (placer14) [inline test/factory/contract.go]: Why can't `fmt` make up it's mind? I hate how this makes our history more noisy.

File: repo/migrations/Migration009.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+	Migration009CreateSalesTable     = "create table sales (orderID text primary key not null, contract blob, state integer, read integer, timestamp integer, total integer, thumbnail text, buyerID text, buyerHandle text, title text, shippingName text, shippingAddress text, paymentAddr text, funded integer, transactions blob, needsSync integer, lastDisputeTimeoutNotifiedAt integer not null default 0);"
+	Migration009CreateSalesIndex     = "create index index_sales on sales (paymentAddr, timestamp);"
+	Migration009CreatePurchasesTable = "create table purchases (orderID text primary key not null, contract blob, state integer, read integer, timestamp integer, total integer, thumbnail text, vendorID text, vendorHandle text, title text, shippingName text, shippingAddress text, paymentAddr text, funded integer, transactions blob, lastDisputeTimeoutNotifiedAt integer not null default 0, lastDisputeExpiryNotifiedAt integer not null default 0, disputedAt integer not null default 0);"
```
Comment (placer14) [inline repo/migrations/Migration009.go]: I like to indicate in the variable name that these schemas are as they existed prior to this migration.

**Review by placer14 — APPROVED** [review]
Some thoughts but nothing that should hold this up. LGTM!

File: repo/db/util.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	paymentCoin := contract.BuyerOrder.Payment.Coin
+	if paymentCoin != "" {
+		return paymentCoin
+	}
+
+	if len(contract.VendorListings[0].Metadata.AcceptedCurrencies) > 0 {
+		paymentCoin = contract.VendorListings[0].Metadata.AcceptedCurrencies[0]
+	}
+
+	return paymentCoin
```
[the remark this answers is not in the document]
Comment (tyler-smith) [reply repo/db/util.go]: VendorListings having a single element is already an invariant. We can make it return an error if there's no listing and fail to create the record if that's preferable.

**Pull request merged**